sound/soc/sof/ipc4-topology.c

```diff
@@ -1116,6 +1132,27 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 
 		pipe_widget = swidget->pipe_widget;
 		pipeline = pipe_widget->private;
+
+		if (pipeline->use_chain_dma) {
+			u32 host_dma_id;
+			u32 fifo_size;
+
+			host_dma_id = platform_params->stream_tag - 1;
+			pipeline->msg.primary &= ~SOF_IPC4_GLB_CHAIN_DMA_HOST_ID_MASK;
+			pipeline->msg.primary |= SOF_IPC4_GLB_CHAIN_DMA_HOST_ID(host_dma_id);
+
+			if (params_width(fe_params) > 16)
+				pipeline->msg.primary &= ~SOF_IPC4_GLB_CHAIN_DMA_SCS_MASK;
+			else
+				pipeline->msg.primary |= SOF_IPC4_GLB_CHAIN_DMA_SCS_MASK;
+
+			fifo_size = 2 * ((params_rate(fe_params) * params_channels(fe_params) *
+					  params_width(fe_params))/ 8000);
```

Comment:
that formula looks like art or black magic. What is the '2' factor for and why would we divide by 8000?

Reply:
Oh, this came from @ranj063, and I only added the sample width to the equation. It gives reasonable nubers for 48kHz stereo 16 and 32-bit (384 and 768 bytes), but now that I look the corresponding fw code I am not sure this would even work for 44.1kHz:
https://github.com/thesofproject/sof/blob/eb712a40db3dae1be73acda772b2e62dd73a92a1/src/ipc/ipc4/helper.c#L499

But what would be better? The size of required buffer should anyway be proportional to sample-rate, sample-width, and number channels, no?

Reply:
> that formula looks like art or black magic. What is the '2' factor for and why would we divide by 8000?

8 is for converting to bytes and 1000 to converto to ms. So the fifo_size should be a minimum of 2ms buffer.

Reply:
shouldn't we use snd_pcm_format_physical_width, i.e. the container size, rather the number of bits valid?

I find it scary that we'd hard-code the tick size. It might be better to have a define or a topology configuration that we make sure if larger than 2ms.

also I don't get why it's exposed as a fifo, this is a DMA buffer isn't it?

Reply:
'fifo_size' is the name used in cSOF Firmware Interface spec.

Reply:
that's worthy of a comment, there's another FIFO in HDaudio but a real-one :-)